Sprint_3

ArtMarenkov opened the pull request:
В 'conftest.py' реализованы фикстуры.
В 'constructor.py' реализованы тесты на переключение разделов в "Конструкторе".
В 'log_in.py'реализованы тесты на вход зарегистрированного пользователя.
В 'log_out.py' реализованы тесты на выход из УЗ пользователя.
В 'personal_account.py'реализован тест на переход в личный кабинет пользователя.
В 'registration.py' реализованы тесты на регистрацию нового пользователя.
В 'transitions_from_pers_acc.py' реализованы тесты на переход из личного кабинета пользователя в "Конструктор".

ArtMarenkov committed:
В 'conftest.py' реализованы фикстуры.
В 'constructor.py' реализованы тесты на переключение разделов в "Конструкторе".
В 'log_in.py'реализованы тесты на вход зарегистрированного пользователя.
В 'log_out.py' реализованы тесты на выход из УЗ пользователя.
В 'personal_account.py'реализован тест на переход в личный кабинет пользователя.
В 'registration.py' реализованы тесты на регистрацию нового пользователя.
В 'transitions_from_pers_acc.py' реализованы тесты на переход из личного кабинета пользователя в "Конструктор".

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,19 @@
+    reg_mail = f'{reg_name}' + '@gmail.com'
+    reg_password ='qwerty' + str(random.randint(1000, 999999999))
+
+    return {"reg_name":reg_name, "reg_mail":reg_mail, "reg_password": reg_password}
```
Можно лучше: тут проще будет вернуть объект data класса, чтобы обращаться было удобнее

ArtMarenkov replied:
А что за data класс? Его создать в отдельном 'data.py'?

AlexKitIt replied:
Ага, лучше так. dataclass это отдельный объект в питоне для хранения статичных данных объединенных логикой использования одновременно. Иными словами, это будет объект User  у него будут атрибуты name, email, password. В метод регистрации мы передадим объект класса, а внутри метода будем обращаться к User.name User.email и т.д. в конкретных полях

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,19 @@
+
+    return {"reg_name":reg_name, "reg_mail":reg_mail, "reg_password": reg_password}
+
+@pytest.fixture
+def login_data():
+
+    login = "marenkov_9@gmail.com"
+    password = "ko16pyth&81"
+
+    return {"login":login, "password": password}
```
Можно лучше: а это в приницпе должен быть обычный класс с данными или отдельные данные а data или над тестовым классом, где они используются. Назначение фикстуры в другом - создание предусловий теста

AlexKitIt commented on tests/constructor.py:
```diff
@@ -0,0 +1,78 @@
+        driver.get("https://stellarburgers.nomoreparties.site/")
+
+        driver.find_element(By.XPATH, "//p[text() = 'Личный Кабинет']").click()
+        WebDriverWait(driver, 2).until(expected_conditions.visibility_of_element_located((By.CSS_SELECTOR, "button[class = 'button_button__33qZ0 button_button_type_primary__1O7Bx button_button_size_medium__3zxIa']")))
```
Нужно исправить: все локаторы должны храниться в отдельном модуле и импортироваться в тесты. Тут их хранить  - сложно поддерживать

AlexKitIt commented on tests/constructor.py:
```diff
@@ -0,0 +1,78 @@
+        login = login_data['login']
+        password = login_data['password']
+
+        driver = webdriver.Chrome()
+        driver.get("https://stellarburgers.nomoreparties.site/")
```
Нудно исправить: это все, как и driver.quit() должно находиться в фикстуре, потому как это и есть не сам тест, а то, что ему предшествует. Кроме того, если assert завалится - инстанс драйвера не закроется, потому как после assert код выполняться не будет

AlexKitIt commented on tests/constructor.py:
```diff
@@ -0,0 +1,78 @@
+        WebDriverWait(driver, 2).until(expected_conditions.visibility_of_element_located((By.XPATH, ".//span[@class = 'text text_type_main-default' and text() = 'Булки']")))
+        driver.find_element(By.XPATH, ".//span[@class = 'text text_type_main-default' and text() = 'Булки']").click()
+
+        assert "tab_tab_type_current" in  driver.find_element(By.XPATH, ".//span[@class = 'text text_type_main-default' and text() = 'Булки']/..").get_attribute('class')
```
Можно лучше: если добавишь assert message, то при разборе будет проще и понятнее почему, а главное что-именно у нас упало

AlexKitIt commented on tests/constructor.py:
```diff
@@ -0,0 +1,78 @@
+        driver.find_element(By.XPATH, "//p[text() = 'Личный Кабинет']").click()
+        WebDriverWait(driver, 2).until(expected_conditions.visibility_of_element_located((By.CSS_SELECTOR, "button[class = 'button_button__33qZ0 button_button_type_primary__1O7Bx button_button_size_medium__3zxIa']")))
+
+        driver.find_element(By.XPATH, ".//input[@class = 'text input__textfield text_type_main-default' and @type = 'text']").send_keys(login)
+        driver.find_element(By.XPATH, ".//input[@class = 'text input__textfield text_type_main-default' and @type = 'password']").send_keys(password)
```
Нужно исправить: локаторы следует писать более осмысленно, привязка ко всем классам таким образом уронит тест, если изменится фронт, т.к. все эти классы используются только для css. Кроме того у конкретных локаторов есть вполне очевидные атрибуты для привязки
Локаторы нужно пересмотреть по возможности все

ArtMarenkov committed:
Внесены исправления по замечаниям.

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,72 @@
+
+@pytest.fixture
+def login_existed():
+    return "marenkov_9@gmail.com"
```
Воу, в таком исполнении прошлый вариант был гораздо лучше. Если на каждую строку писать по отдельному методу - тут жесть что начнется

ArtMarenkov committed:
1) удалён 'random.py';
2) создан 'data.py' в папке data, где реализован dataclass User;
3) из 'conftest.py' убраны фикстуры, связанные с даннымии пользователя;
4) проапдейчены 'log_in.py' и 'registration.py' в связи с изменениями выше.

AlexKitIt commented on data/data.py:
```diff
@@ -0,0 +1,23 @@
+    user_email: str = 'marenkov_9@gmail.com'
+    user_password: str = 'ko16pyth&81'
+
+    def __init__(self, name, email, password):
+        self.user_name = name
+        self.user_email = email
+        self.user_password = password
```
Можно лучше: для data классов не требуется init, он будет интерпритирован в моменте его встречи в коде еще до запуска основного алгоритма
